clippy_lints/src/indexing_slicing.rs

```diff
@@ -111,17 +111,43 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for IndexingSlicing {
                 // Ranged indexes, i.e. &x[n..m], &x[n..], &x[..n] and &x[..]
                 if let ty::Array(_, s) = ty.sty {
                     let size: u128 = s.assert_usize(cx.tcx).unwrap().into();
-                    // Index is a constant range.
-                    if let Some((start, end)) = to_const_range(cx, range, size) {
-                        if start > size || end > size {
-                            utils::span_lint(
-                                cx,
-                                OUT_OF_BOUNDS_INDEXING,
-                                expr.span,
-                                "range is out of bounds",
-                            );
-                        }
-                        return;
+
+                    match to_const_range(cx, range, size) {
+                        (None, None) => {},
+                        (Some(start), None) => {
```

Comment:
I think you could deduplicate these calls by first checking the start for `Some` and if that + the `> size` check fails, check the end for `Some`. That way you only have two places that emit the lint and they have a targetted text for the start/end of the range